File: src/main/java/bad/robot/radiate/teamcity/Username.java
package bad.robot.radiate.teamcity;

@Deprecated
public class Username extends bad.robot.http.configuration.AbstractValueType<String> {

    public static Username username(String username) {
        if (username == null)
            return new NoUsername();
        return new Username(username);
    }

    protected Username(String value) {
        super(value);
    }

    bad.robot.http.configuration.Username asSimpleHttp() {
        return bad.robot.http.configuration.Username.username(value);
    }

}

File: src/main/scala/bad/robot/radiate/teamcity/Username.scala
package bad.robot.radiate.teamcity

import bad.robot.radiate.OptionSyntax._

import scalaz.{Success, Validation}
import scalaz.syntax.std.option._

object Username {

  def validate(username: Option[String]): Validation[String, Username] = {
    Success(NonEmptyOption(username).map(Username.apply).getOrElse(NoUsername))
  }

  // deprecated
  def username(username: String): Username = {
    if (username == null) NoUsername else new Username(username)
  }
}

// todo implicit to convert the SimpleHTTP username
case class Username (value: String) {
  private[teamcity] def asSimpleHttp: bad.robot.http.configuration.Username = {
    bad.robot.http.configuration.Username.username(value)
  }
}

object NoUsername extends Username("no username")